clients/drcachesim/tools/filter/record_filter.cpp

```diff
@@ -445,14 +549,23 @@ record_filter_t::process_chunk_encodings(per_shard_t *per_shard, trace_entry_t &
         // encodings at chunk boundaries regardless.
         if (TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, per_shard->filetype) &&
             per_shard->cur_chunk_pcs.find(entry.addr) == per_shard->cur_chunk_pcs.end()) {
-            if (per_shard->pc2encoding.find(entry.addr) == per_shard->pc2encoding.end()) {
+            if (per_shard->per_input == nullptr)
+                return "Invalid input id for instruction";
+            std::lock_guard<std::mutex> guard(per_shard->per_input->lock);
+            if (per_shard->per_input->pc2encoding.find(entry.addr) ==
+                per_shard->per_input->pc2encoding.end()) {
                 return "Missing encoding for PC " + std::to_string(entry.addr) +
+                    " in shard " +
+                    std::to_string(per_shard->shard_stream->get_shard_index()) +
+                    " for input " +
+                    std::to_string(per_shard->shard_stream->get_input_id()) +
                     " at input entry " + std::to_string(per_shard->input_entry_count);
             }
             VPRINT(this, 3,
                    "output new-chunk encoding chunk=%" PRIu64 " ref=%" PRIu64 "\n",
                    per_shard->chunk_ordinal, per_shard->cur_refs);
-            if (!write_trace_entries(per_shard, per_shard->pc2encoding[entry.addr])) {
+            if (!write_trace_entries(per_shard,
+                                     per_shard->per_input->pc2encoding[entry.addr])) {
```

Comment:
I'm not sure there's a problem.  I have in my notes a TODO to think about JIT complications but I believe it works out, If each input thread has all its encodings correct.  If the scheduler re-orders thread B to execute a newer instr before thread A writes it: that's a dynamic schedule problem where the scheduler fails to honor inter-thread deps, not a record filter problem.

Reply:
> that's a dynamic schedule problem where the scheduler fails to honor inter-thread deps

Okay so that's an inter thread dependency that the scheduler will honor.